Tests/CommandsTests/TestToolTests.swift

```diff
@@ -11,13 +11,30 @@
 import XCTest
 
 import TestSupport
+import Basic
 import Commands
 
 final class TestToolTests: XCTestCase {
-    private func execute(_ args: [String]) throws -> String {
-        return try SwiftPMProduct.SwiftTest.execute(args, printIfError: true)
+    private func execute(_ args: [String], packagePath: AbsolutePath? = nil) throws -> String {
+        return try SwiftPMProduct.SwiftTest.execute(args, packagePath: packagePath, printIfError: true)
     }
-    
+
+    private enum BadCodeException : Error {
+        case executionFailure
+    }
+
+    private func executeFailing(_ args: [String], packagePath: AbsolutePath? = nil) throws {
+        do {
+            // The following is expected to fail.
+            _ = try SwiftPMProduct.SwiftRun.execute(args, packagePath: packagePath, printIfError: false)
+        } catch SwiftPMProductError.executionFailure {
```

Comment:
I don't think we require another layer of error and this utility method. We should be able to use this error case to do the assertion.

Reply:
We can't do that, since for XCTAssertThrows() we need a concrete, specific error object and can't use something like a `WhateverExitCode(let exitCode)`.

Reply:
We can do two things here:

1. Conform `SwiftPMProductError` to `Equatable`. This will allow using `XCTAssertThrows` on executionFailure.
2. Add a new helper method that doesn't care about enums being `Equatable`. Something like this:

```swift
public func XCTAssertThrows<T: Swift.Error, U>(
    _ expression: @autoclosure () throws -> U,
    file: StaticString = #file,
    line: UInt = #line,
    _ body: (T) -> Void
) {
    do {
        _ = try expression()
        XCTFail("body completed successfully", file: file, line: line)
    } catch let error as T {
        body(error)
    } catch {
        XCTFail("unexpected error thrown", file: file, line: line)
    }
}

func testSanitizeThreadAddress() throws {
    fixture(name: "Miscellaneous/SanitizersTest") { path in
        let expression = {
            try SwiftPMProduct.SwiftRun.execute(
                ["--sanitize=thread", "--sanitize=address"], packagePath: path, printIfError: false
            )
        }
        XCTAssertThrows(expression) { (error: SwiftPMProductError) in
            switch error {
            case .executionFailure:
                return
            default: XCTFail("Unexpected error \(error)")
            }
        }
    }
}
```

Reply:
I'd avoid introducing the high level (library-level) things for one-off tasks. It seems that both elevating SwiftPMProductError to Equatable and adding a special XCTAssertThrows are of that kind: premature abstraction.

Reply:
In any case, please pick some specific thing you'd want me to do and I'll add it.

Reply:
Yeah, I am not so sure about 1 but 2 is definitely useful. There are several places in the current test code that can benefit from 2. The new method can be placed in `Sources/TestSupport/XCTAssertHelpers.swift`.

Also, if you think something is a bad idea, we should discuss more, I am always open to discuss ideas. The main reason I don't like the current approach is that it adds an additional layer of indirection, which someone might need to dig through in future.

Reply:
Done